### PR TITLE
Add initial support for promises

### PR DESCRIPTION
**Pull request by aaronmunsters**
It adds the following:

- A job queue (in `Context`)
- The constructor [`Promise`](https://tc39.es/ecma262/#sec-promise-executor)
- [`Promise.prototype.then`](https://tc39.es/ecma262/#sec-promise.prototype.then)
- The additional needed infrastructure
  - [`PerformPromiseThen ( promise, onFulfilled, onRejected [ , resultCapability ] )`](https://tc39.es/ecma262/#sec-performpromisethen)
  - [`TriggerPromiseReactions ( reactions, argument )`](https://tc39.es/ecma262/#sec-triggerpromisereactions)
  - [`RejectPromise ( promise, reason )`](https://tc39.es/ecma262/#sec-rejectpromise)
  - [`FulfillPromise ( promise, value )`](https://tc39.es/ecma262/#sec-fulfillpromise)
  - [`CreateResolvingFunctions ( promise )`](https://tc39.es/ecma262/#sec-createresolvingfunctions)
  - [`NewPromiseCapability ( C )`](https://tc39.es/ecma262/#sec-newpromisecapability)
  - [`NewPromiseReactionJob ( reaction, argument )`](https://tc39.es/ecma262/#sec-newpromisereactionjob)
  - [`NewPromiseResolveThenableJob ( promiseToResolve, thenable, then )`](https://tc39.es/ecma262/#sec-newpromiseresolvethenablejob)
- A test case showcasing the run-to-completion semantics.

An example program that shows the control flow with this addition is:
```javascript
new Promise((res, rej) => {
  console.log("A");
  res(undefined);
}).then((_) => console.log("B"));
console.log("C");
```
Which would output:
```
A
C
B
```

Feedback is most welcome! :wink:

### CODE REVIEW

**Review by jedel1043 — CHANGES_REQUESTED** [review]
Very good work! I really like the idea of having a simple queue executor until we make the VM async.
This is a superficial review, I'll do a proper review later when I'm free :)

**Comment by Razican** [comment]
#### VM implementation
[7](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:7)
  | Test result | main count | PR count | difference |
[8](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:8)
  | :---------: | :----------: | :------: | :--------: |
[9](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:9)
  | Total | 88,4[10](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:10) | 88,410 | 0 |
10
  | Passed | 43,978 | 44,104 | **+126** |
[11](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:11)
  | Ignored | 21,481 | 21,481 | 0 |
[12](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:12)
  | Failed | 22,951 | 22,825 | **-126** |
[13](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:13)
  | Panics | 0 | 2 | **+2** |
[14](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:14)
  | Conformance | 49.74% | 49.89% | **+0.14%** |
[15](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:15)
  
[16](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:16)
  <details><summary><b>Fixed tests (126):</b></summary>
[17](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:17)
  
[18](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:18)
  ```
[19](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:19)
  test/built-ins/Object/seal/seal-promise.js [strict mode] (previously Failed)
[20](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:20)
  test/built-ins/Object/seal/seal-promise.js (previously Failed)
[21](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:21)
  test/built-ins/Promise/resolve-function-length.js [strict mode] (previously Failed)
[22](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:22)
  test/built-ins/Promise/resolve-function-length.js (previously Failed)
[23](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:23)
  test/built-ins/Promise/constructor.js [strict mode] (previously Failed)
[24](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:24)
  test/built-ins/Promise/constructor.js (previously Failed)
[25](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:25)
  test/built-ins/Promise/resolve-function-extensible.js [strict mode] (previously Failed)
[26](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:26)
  test/built-ins/Promise/resolve-function-extensible.js (previously Failed)
[27](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:27)
  test/built-ins/Promise/executor-not-callable.js [strict mode] (previously Failed)
[28](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:28)
  test/built-ins/Promise/executor-not-callable.js (previously Failed)
[29](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:29)
  test/built-ins/Promise/reject-function-prototype.js [strict mode] (previously Failed)
[30](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:30)
  test/built-ins/Promise/reject-function-prototype.js (previously Failed)
[31](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:31)
  test/built-ins/Promise/get-prototype-abrupt.js [strict mode] (previously Failed)
[32](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:32)
  test/built-ins/Promise/get-prototype-abrupt.js (previously Failed)
[33](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:33)
  test/built-ins/Promise/reject-function-length.js [strict mode] (previously Failed)
[34](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:34)
  test/built-ins/Promise/reject-function-length.js (previously Failed)
[35](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:35)
  test/built-ins/Promise/get-prototype-abrupt-executor-not-callable.js [strict mode] (previously Failed)
[36](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:36)
  test/built-ins/Promise/get-prototype-abrupt-executor-not-callable.js (previously Failed)
[37](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:37)
  test/built-ins/Promise/resolve-function-nonconstructor.js [strict mode] (previously Failed)
[38](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:38)
  test/built-ins/Promise/resolve-function-nonconstructor.js (previously Failed)
[39](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:39)
  test/built-ins/Promise/executor-call-context-sloppy.js (previously Failed)
[40](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:40)
  test/built-ins/Promise/name.js [strict mode] (previously Failed)
[41](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:41)
  test/built-ins/Promise/name.js (previously Failed)
[42](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:42)
  test/built-ins/Promise/resolve-function-property-order.js [strict mode] (previously Failed)
[43](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:43)
  test/built-ins/Promise/resolve-function-property-order.js (previously Failed)
[44](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:44)
  test/built-ins/Promise/is-a-constructor.js [strict mode] (previously Failed)
[45](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:45)
  test/built-ins/Promise/is-a-constructor.js (previously Failed)
[46](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:46)
  test/built-ins/Promise/undefined-newtarget.js [strict mode] (previously Failed)
[47](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:47)
  test/built-ins/Promise/undefined-newtarget.js (previously Failed)
[48](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:48)
  test/built-ins/Promise/property-order.js [strict mode] (previously Failed)
[49](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:49)
  test/built-ins/Promise/property-order.js (previously Failed)
[50](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:50)
  test/built-ins/Promise/reject-function-property-order.js [strict mode] (previously Failed)
[101](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:101)
  test/built-ins/Promise/all/S25.4.4.1_A4.1_T1.js [strict mode] (previously Failed)
[102](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:102)
  test/built-ins/Promise/all/S25.4.4.1_A4.1_T1.js (previously Failed)
[103](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:103)
  test/built-ins/Promise/all/ctx-non-object.js [strict mode] (previously Failed)
[104](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:104)
  test/built-ins/Promise/all/ctx-non-object.js (previously Failed)
[105](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:105)
  test/built-ins/Promise/all/not-a-constructor.js [strict mode] (previously Failed)
[106](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:106)
  test/built-ins/Promise/all/not-a-constructor.js (previously Failed)
[107](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:107)
  test/built-ins/Promise/reject/ctx-non-ctor.js [strict mode] (previously Failed)
[108](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:108)
  test/built-ins/Promise/reject/ctx-non-ctor.js (previously Failed)
[109](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:109)
  test/built-ins/Promise/reject/ctx-non-object.js [strict mode] (previously Failed)
[110](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:110)
  test/built-ins/Promise/reject/ctx-non-object.js (previously Failed)
[111](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:111)
  test/built-ins/Promise/reject/not-a-constructor.js [strict mode] (previously Failed)
[112](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:112)
  test/built-ins/Promise/reject/not-a-constructor.js (previously Failed)
[113](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:113)
  test/built-ins/Promise/reject/S25.4.4.4_A3.1_T1.js [strict mode] (previously Failed)
[114](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:114)
  test/built-ins/Promise/reject/S25.4.4.4_A3.1_T1.js (previously Failed)
[115](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:115)
  test/built-ins/Promise/prototype/S25.4.5_A3.1_T1.js [strict mode] (previously Failed)
[116](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:116)
  test/built-ins/Promise/prototype/S25.4.5_A3.1_T1.js (previously Failed)
[117](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:117)
  test/built-ins/Promise/prototype/prop-desc.js [strict mode] (previously Failed)
[118](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:118)
  test/built-ins/Promise/prototype/prop-desc.js (previously Failed)
[119](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:119)
  test/built-ins/Promise/prototype/S25.4.4.2_A1.1_T1.js [strict mode] (previously Failed)
[120](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:120)
  test/built-ins/Promise/prototype/S25.4.4.2_A1.1_T1.js (previously Failed)
[121](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:121)
  test/built-ins/Promise/prototype/proto.js [strict mode] (previously Failed)
[122](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:122)
  test/built-ins/Promise/prototype/proto.js (previously Failed)
[123](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:123)
  test/built-ins/Promise/prototype/then/S25.4.5.3_A2.1_T2.js [strict mode] (previously Failed)
[124](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:124)
  test/built-ins/Promise/prototype/then/S25.4.5.3_A2.1_T2.js (previously Failed)
[125](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:125)
  test/built-ins/Promise/prototype/then/prop-desc.js [strict mode] (previously Failed)
[126](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:126)
  test/built-ins/Promise/prototype/then/prop-desc.js (previously Failed)
[127](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:127)
  test/built-ins/Promise/prototype/then/not-a-constructor.js [strict mode] (previously Failed)
[128](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:128)
  test/built-ins/Promise/prototype/then/not-a-constructor.js (previously Failed)
[129](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:129)
  test/built-ins/Promise/prototype/then/name.js [strict mode] (previously Failed)
[130](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:130)
  test/built-ins/Promise/prototype/then/name.js (previously Failed)
[131](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:131)
  test/built-ins/Promise/prototype/then/ctor-null.js [strict mode] (previously Failed)
[132](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:132)
  test/built-ins/Promise/prototype/then/ctor-null.js (previously Failed)
[133](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:133)
  test/built-ins/Promise/prototype/then/S25.4.5.3_A2.1_T1.js [strict mode] (previously Failed)
[134](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:134)
  test/built-ins/Promise/prototype/then/S25.4.5.3_A2.1_T1.js (previously Failed)
[135](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:135)
  test/built-ins/Promise/prototype/then/S25.4.5.3_A1.1_T1.js [strict mode] (previously Failed)
[136](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:136)
  test/built-ins/Promise/prototype/then/S25.4.5.3_A1.1_T1.js (previously Failed)
[137](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:137)
  test/built-ins/Promise/prototype/finally/not-a-constructor.js [strict mode] (previously Failed)
[138](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:138)
  test/built-ins/Promise/prototype/finally/not-a-constructor.js (previously Failed)
[139](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:139)
  test/built-ins/Promise/prototype/catch/not-a-constructor.js [strict mode] (previously Failed)
[140](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:140)
  test/built-ins/Promise/prototype/catch/not-a-constructor.js (previously Failed)
[141](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:141)
  test/built-ins/Promise/prototype/catch/this-value-non-object.js [strict mode] (previously Failed)
[142](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:142)
  test/built-ins/Promise/prototype/catch/this-value-non-object.js (previously Failed)
[143](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:143)
  test/built-ins/Promise/prototype/catch/this-value-then-not-callable.js [strict mode] (previously Failed)
[144](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:144)
  test/built-ins/Promise/prototype/catch/this-value-then-not-callable.js (previously Failed)
[145](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:145)
  ```
[146](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:146)
  </details>
[147](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:147)
  
[148](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:148)
  <details><summary><b>New panics (2):</b></summary>
[149](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:149)
  
[150](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:150)
  ```
[151](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:151)
  test/built-ins/Promise/prototype/no-promise-state.js [strict mode] (previously Failed)
[152](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:152)
  test/built-ins/Promise/prototype/no-promise-state.js (previously Failed)
[153](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:153)
  ```
[154](https://github.com/boa-dev/boa/runs/5528502927?check_suite_focus=true#step:11:154)
  </details>

**Comment by Razican** [comment]
I really like how this looks! It's a great step in the right direction. Maybe you can already remove the ignored flag:async tests from the test_ignore.txtI still have to review this properly (even though most of those require extra tester support).

In any case, there is something that has already caught my attention. The usage of the [queues](https://crates.io/crates/queues) crate. This crate was last updated in 2018 (more than 3 years ago), so I'm not comfortable adding it as a dependency. In fact, I might even prefer to create a "boa_queues" crate or something and adapt it. The crate itself is only 1 Rust file (we need to check licensing, though).

File: boa_engine/src/builtins/promise/mod.rs [hunk 1]
```diff
@@ -0,0 +1,731 @@
...
+struct ResolvingFunctionsRecord {
+    resolve: JsValue,
+    reject: JsValue,
+}
+
+#[derive(Debug, Trace, Finalize)]
+struct RejectResolveCaptures {
+    promise: JsObject,
+    already_resolved: JsObject,
+}
```
Comment (Razican) [inline boa_engine/src/builtins/promise/mod.rs:206]: Could we add some documentation to these structures? Also, they must all implement `Debug`, `Clone` and `Copy`, when possible

File: boa_engine/src/builtins/promise/mod.rs [hunk 2]
```diff
@@ -0,0 +1,731 @@
...
+    Context, JsResult,
+};
+
+use self::promise_job::PromiseJob;
+
+use super::JsArgs;
+
+/// JavaScript `Array` built-in implementation.
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct Array;
```
Comment (Razican) [inline boa_engine/src/builtins/promise/mod.rs:33]: Why is this needed here?

File: boa_engine/src/builtins/promise/mod.rs [hunk 3]
```diff
@@ -0,0 +1,731 @@
...
+struct PromiseCapability {
+    promise: JsValue,
+    resolve: JsValue,
+    reject: JsValue,
+}
+
+#[derive(Debug, Trace, Finalize)]
+struct PromiseCapabilityCaptures {
+    promise_capability: Gc<boa_gc::Cell<PromiseCapability>>,
+}
```
Comment (Razican) [inline boa_engine/src/builtins/promise/mod.rs:74]: I would like to have some documentation for these structures, to understand their purpose, if there is any reference to them in the spec, and `ReactionType` and `PromiseState` can probably implement `Copy`, which would make things easier.

For `Copy` types, you will probably need to implement `Trace` and `Finalize` separately, with an empty implementation, or better: whenever a structure requires them in, use the `#[unsafe_ignore_trace]` attribute on the field.

File: boa_engine/src/builtins/promise/mod.rs [hunk 4]
```diff
@@ -0,0 +1,731 @@
+//! This module implements the global `Promise` object.
+
+#![allow(dead_code, unused_results, unused_variables)]
```
Comment (Razican) [inline boa_engine/src/builtins/promise/mod.rs:3]: I think this shouldn't stay here, right? We don't want to silence warnings :)

File: boa_engine/src/builtins/promise/promise_job.rs [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+use crate::{
+    builtins::promise::{ReactionRecord, ReactionType},
+    job::JobCallback,
+    object::{FunctionBuilder, JsObject},
+    Context, JsValue,
+};
+
+use super::{Promise, PromiseCapability};
+
+pub(crate) struct PromiseJob;
```
Comment (Razican) [inline boa_engine/src/builtins/promise/promise_job.rs:12]: Could we add some documentation here, add the reference to the spec, and derive `Debug`, `Clone` and `Copy` for it?

File: boa_engine/src/builtins/promise/promise_job.rs [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+
+#[derive(Debug, Trace, Finalize)]
+struct ReactionJobCaptures {
+    reaction: ReactionRecord,
+    argument: JsValue,
+}
```
Comment (Razican) [inline boa_engine/src/builtins/promise/promise_job.rs:18]: This would need some documentation to explain where does it come from, what it's used for, how...

File: boa_engine/src/builtins/promise/promise_job.rs [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+
+impl JobCapture {
+    fn new(promise_to_resolve: JsObject, thenable: JsValue, then: JobCallback) -> Self {
+        Self {
+            promise_to_resolve,
+            thenable,
+            then,
+        }
+    }
+}
```
Comment (Razican) [inline boa_engine/src/builtins/promise/promise_job.rs:194]: This would need some documentation to know where it's coming from, how to use it, what it means and so on. Maybe even a spec link. Also for the `new()` function.

File: boa_engine/src/job.rs [hunk 8]
```diff
@@ -0,0 +1,39 @@
+use crate::{Context, JsResult, JsValue};
+
+use gc::{Finalize, Trace};
+
+#[derive(Debug, Clone, Trace, Finalize)]
+pub struct JobCallback {
```
Comment (Razican) [inline boa_engine/src/job.rs:6]: I would like to have a bit of documentation in this structure, to understand it better.

File: boa_engine/src/job.rs [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    pub fn run(&self, context: &mut Context) {
+        let callback = match *self.callback {
+            JsValue::Object(ref object) if object.is_callable() => object.clone(),
+            _ => panic!("Callback is not a callable object"),
+        };
+
+        let _callback_result = callback.__call__(&JsValue::Undefined, &[], context);
+    }
```
Comment (Razican) [inline boa_engine/src/job.rs:38]: Could we add documentation to these functions?

**Review by Razican — CHANGES_REQUESTED** [review]
Hi, I got the time to review it a bit more, and I noticed it's missing some documentation. The rest looks pretty good!

Also, we must check where are those panics coming from and fix them.

**Comment by lastmjs** [comment]
Does this help us get toward allowing https://boa-dev.github.io/boa/doc/boa_engine/builtins/function/type.NativeFunctionSignature.html to be asynchronous?

**Comment by aaronmunsters** [comment]
> Does this help us get toward allowing https://boa-dev.github.io/boa/doc/boa_engine/builtins/function/type.NativeFunctionSignature.html to be asynchronous?

This contribution does not add an executor/async rust code,  all code that you would add as native functions will be executed on the single/main rust thread that executes all JS code in order of the jobqueue after executing the main program (run-to-completion).

Say that you would now add `fetch` to Boa which would add a promise that represents the completion of a network request.
If you would now write a program that contains two fetch expressions (each including a `.then(logResult)` statement), then as soon as the main JS code is evaluated, the VM would now process the job queue _in order_ and it would block separately on each network request **(single tread, even the VM)**, rather than performing the networking in parallel and performing the callbacks as soon as one request is processed.

I think it helps towards async builtin functions as it expands upon the ECMAScript standards but that would require making the VM async.

**Comment by lastmjs** [comment]
Even if it is still single-threaded run-to-completion, could we change the NativeFunctionSignature to allow a function that returns a Future? Even if it's blocking under-the-hood, would it still work?

So for example imagine creating a custom Rust function (that does an http request or something) that returns a Future exposed to boa as a function that returns a Promise in JS. Even if it's technically synchronous, this could be very useful for my use case.

File: boa_engine/src/builtins/promise/mod.rs (hunk 2 above)
Comment (aaronmunsters) [reply boa_engine/src/builtins/promise/mod.rs:33]: It is not, thanks for pointing this out :)

File: boa_engine/src/builtins/promise/mod.rs (hunk 4 above)
Comment (aaronmunsters) [reply boa_engine/src/builtins/promise/mod.rs:3]: You're absolutely right, forgot to remove this :)

**Comment by Razican** [comment]
> Even if it is still single-threaded run-to-completion, could we change the NativeFunctionSignature to allow a function that returns a Future? Even if it's blocking under-the-hood, would it still work?
> 
> So for example imagine creating a custom Rust function (that does an http request or something) that returns a Future exposed to boa as a function that returns a Promise in JS. Even if it's technically synchronous, this could be very useful for my use case.

We should manage expectations here. It's important to understand that ECMAScript is single-threaded. Even its async/await is single-threaded always. This requires some guarantees (such as types not implementing `Send`) that collide with the Rust asynchronous runtime (everything must implement `Send`, and it's always treated as multithreaded, even with one thread).

So, `NativeFunction` will probably stay the same, but there might be an option at some point in the future to be able to return something like a `NativePromise`, which could be a `JsValue`. But this requires a lot of thought, and I'm not sure if it's even possible.

What is sure is that while we don't have asynchronous JavaScript, we cannot have asynchronous Rust. And even if/when implement asynchronous Rust, synchronous JavaScript and Rust must still be supported. So, this is a huge step towards enabling asynchronous JavaScript to run in Boa, and this could potentially open the doors to use asynchronous Rust native functions at some point, but it will be very, very tricky.

For now, you can use `block_on()` functions to call asynchronous Rust from a `NativeFunction`, but you will have to handle sending information between threads.

**Comment by lastmjs** [comment]
Thank you very much for the in-depth explanation. Unfortunately I am working in a constrained Wasm environment where block_on does not work, hence my great desire for asynchronus JS and Rust (since it blocks so many capabilities I am trying to give my users). Thanks!

**Comment by jasonwilliams** [comment]
Hopefully i can take a proper look at this soon

**Comment by Razican** [comment]
@aaronmunsters did you have the chance to look at the review comments?

**Comment by aaronmunsters** [comment]
Hi there, unfortunately not yet. This week I've got another deadline coming up, but I'll get at it starting next week 😊

**Review by jasonwilliams — COMMENTED** [review]
Looks like a great start to me. I agree it will need some tidying up from the comments. Job itself will need some documentation describing what it is and how it's used etc etc.

I like the test showing the value after completion.
Having something like [mio](https://docs.rs/mio/0.5.1/mio/struct.EventLoop.html) handle our queue in future for tasks would be nice. 
I think V8 doesn't actually deal with any of that, and Node just adds that logic on, i wonder if we'll eventually need a flag like we had with console.

File: boa_engine/src/context/mod.rs [hunk 10]
```diff
@@ -709,10 +714,21 @@ impl Context {
         self.realm.set_global_binding_number();
         let result = self.run();
         self.vm.pop_frame();
+        self.run_queued_jobs();
```
Comment (jasonwilliams) [inline boa_engine/src/context/mod.rs:717]: one thing that confused me, is in your example i would have expected `count` to be three. Because execute waits here for the promise job to finish before finishing. But it seems that's not the case.  It looks like it returns 2 and you need to call it again (looking at your test).

Can someone explain that?

**Comment by sccgithub** [comment]
Whether there is progress in this feature, it is a very practical feature, I am looking forward to its completion.

File: boa_engine/src/builtins/promise/mod.rs (hunk 1 above)
Comment (jedel1043) [inline boa_engine/src/builtins/promise/mod.rs:206]: You can move this inside `create_resolving_functions`. This should limit its scope, since it's only used internally, and it would make it renameable to only `Captures`.

File: boa_engine/src/builtins/promise/mod.rs [hunk 11]
```diff
@@ -0,0 +1,731 @@
...
+
+struct ResolvingFunctionsRecord {
+    resolve: JsValue,
+    reject: JsValue,
+}
+
+#[derive(Debug, Trace, Finalize)]
+struct RejectResolveCaptures {
+    promise: JsObject,
+    already_resolved: JsObject,
```
Comment (jedel1043) [inline boa_engine/src/builtins/promise/mod.rs:205]: I'm pretty sure you can make `already_resolved` a `bool`.

File: boa_engine/src/builtins/promise/mod.rs [hunk 12]
```diff
@@ -0,0 +1,731 @@
...
+        .name("")
+        .length(1)
+        .constructor(false)
+        .build();
+
+        // 5. Set resolve.[[Promise]] to promise.
+        resolve.set("Promise", promise.clone(), true, context)?;
+
+        // 6. Set resolve.[[AlreadyResolved]] to alreadyResolved.
+        resolve.set("AlreadyResolved", already_resolved.clone(), true, context)?;
```
Comment (jedel1043) [inline boa_engine/src/builtins/promise/mod.rs:423]: This initialization is unnecessary. The spec asks you to set `[[Promise]]` and `[[AlreadyResolved]]`, but those are the captures of `resolve`, which are initialized above. Also applies to the `reject` closure.

File: boa_engine/src/builtins/promise/mod.rs [hunk 13]
```diff
@@ -0,0 +1,731 @@
...
+                        //   b. Return undefined.
+                        return Ok(JsValue::Undefined);
+                    }
+                    Ok(then) => then,
+                };
+
+                // 11. Let thenAction be then.[[Value]].
+                let then_action = then
+                    .as_object()
+                    .expect("rsolution.[[then]] should be an object")
```
Comment (jedel1043) [inline boa_engine/src/builtins/promise/mod.rs:379]:
```suggestion
                    .expect("resolution.[[then]] should be an object")
```

File: boa_engine/src/builtins/promise/mod.rs [hunk 14]
```diff
@@ -0,0 +1,731 @@
...
+    /// More information:
+    ///  - [ECMAScript reference][spec]
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-fulfillpromise
+    pub fn fulfill(&mut self, value: &JsValue, context: &mut Context) -> JsResult<()> {
+        // 1. Assert: The value of promise.[[PromiseState]] is pending.
+        match self.promise_state {
+            PromiseState::Pending => (),
+            _ => return context.throw_error("Expected promise.[[PromiseState]] to be pending"),
+        }
```
Comment (jedel1043) [inline boa_engine/src/builtins/promise/mod.rs:498]: An assertion should be asserted with `assert!` or `debug_assert!`  which more clearly indicates that it's an engine error and not an user error.

**Review by jedel1043 — CHANGES_REQUESTED** [review]
Sorry for the wait! I finally did an extensive review, and there are some details that should improve the semantics and readability of the code.

File: boa_engine/src/context/mod.rs (hunk 10 above)
Comment (jedel1043) [reply boa_engine/src/context/mod.rs:717]: The `Promise` constructor executes its code on creation, and `then` is queued on the job queue. After executing all the statements of the script, `count === 2` and `then` is executed, making `count === 3`. However, the script already returned `2` as the evaluation result, since `let result = self.run()` executes before `self.run_queued_jobs()`. When the next execution occurs, `count` is already 3, so the evaluation returns 3.

**Comment by Razican** [comment]
@aaronmunsters did you have the chance to look at this?

File: boa_engine/src/job.rs [hunk 15]
```diff
@@ -0,0 +1,39 @@
+use crate::{Context, JsResult, JsValue};
+
+use gc::{Finalize, Trace};
+
+#[derive(Debug, Clone, Trace, Finalize)]
+pub struct JobCallback {
+    callback: Box<JsValue>,
```
[the remark this answers is not in the document]
Comment (Razican) [reply boa_engine/src/job.rs:7]: Done in #2107.

File: boa_engine/src/context/mod.rs [hunk 16]
```diff
@@ -709,10 +714,21 @@ impl Context {
...
 
+    fn run_queued_jobs(&mut self) {
+        while self.promise_job_queue.size() != 0 {
+            let job = self
+                .promise_job_queue
+                .remove()
+                .expect("Job Queue should not be empty");
+            job.run(self);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Razican) [reply boa_engine/src/context/mod.rs:730]: Done in #2107.

File: boa_engine/src/builtins/promise/mod.rs (hunk 2 above)
Comment (Razican) [reply boa_engine/src/builtins/promise/mod.rs:33]: Added the comment to #2107

File: boa_engine/src/context/mod.rs [hunk 17]
```diff
@@ -86,6 +88,8 @@ pub struct Context {
     strict: bool,
 
     pub(crate) vm: Vm,
+
+    pub(crate) promise_job_queue: Queue<Box<JobCallback>>,
```
[the remark this answers is not in the document]
Comment (Razican) [reply boa_engine/src/context/mod.rs:92]: This was changed by a `VecDeque` in #2107, I thought a growable double-ended ring buffer could be more performant. I also removed the `Box` indirection (and the second indirection in `JobCallback`). Feel free to test if a `Vec` is faster.

File: boa_engine/src/builtins/promise/mod.rs (hunk 14 above)
Comment (Razican) [reply boa_engine/src/builtins/promise/mod.rs:498]: This has been fixed in #2107.

File: boa_engine/src/builtins/promise/mod.rs (hunk 12 above)
Comment (Razican) [reply boa_engine/src/builtins/promise/mod.rs:423]: This has been fixed in #2107.

File: boa_engine/src/builtins/promise/mod.rs (hunk 1 above)
Comment (Razican) [reply boa_engine/src/builtins/promise/mod.rs:206]: This is also being used in job.rs, so I left it as-is in #2107. Feel free to review it there.
